---
layout:     post
catalog: true
title:      ‘Simulating genetic data with R： an example with deleterious variants (and a pun)’
subtitle:      转载自：http://feedproxy.google.com/~r/RBloggers/~3/xA52VRspUm4/
date:      2019-06-16
author:      mrtnj
tags:
    - simparam
    - breeding
    - breeds
    - simulations
    - distribution
---





A few weeks ago, I gave a talk at the Edinburgh R users group EdinbR on the RAGE paper. Since this is an R meetup, the talk concentrated on the mechanics of genetic data simulation and with the paper as a case study. I showed off some of what Chris Gaynor’s AlphaSimR can do, and how we built on that to make the specifics of this simulation study. The slides are on the EdinbR Github.

Genetic simulation is useful for all kinds of things. Sure, they’re only as good as the theory that underpins them, but the willingness to try things out in simulations is one of the things I always liked about breeding research.

This is my description of the logic of genetic simulation: we think of the genome as a large table of genotypes, drawn from some distribution of allele frequencies.

![](https://martinsbioblogg.files.wordpress.com/2019/06/slide04.png?w=456&is-pending-load=1)
![](https://martinsbioblogg.files.wordpress.com/2019/06/slide04.png?w=456)


To make an utterly minimal simulation, we could draw allele frequencies from some distribution (like a Beta distribution), and then draw the genotypes from a binomial distribution. Done! 

![](https://martinsbioblogg.files.wordpress.com/2019/06/slide05.png?w=640&is-pending-load=1#038;h=359&fit=456%2C456)
![](https://martinsbioblogg.files.wordpress.com/2019/06/slide05.png?w=640&h=359&fit=456%2C456)


However, there is a ton of nuance we would like to have: chromosomes, linkage between variants, sexes, mating, selection …

![](https://martinsbioblogg.files.wordpress.com/2019/06/slide06.png?w=640&is-pending-load=1#038;h=359&fit=456%2C456)
![](https://martinsbioblogg.files.wordpress.com/2019/06/slide06.png?w=640&h=359&fit=456%2C456)


AlphaSimR addresses all of this, and allows you to throw individuals and populations around to build pretty complicated designs. Here is the small example simulation I used in the talk.

This code builds a small livestock population, breeds it for ten generations, and looks at the resulting selection response in the form of a shift of the genetic mean, and the changes in the underlying distribution of causative variants. Here are the resulting plots:

![](https://martinsbioblogg.files.wordpress.com/2019/06/blogpost.png?w=450&is-pending-load=1#038;h=264&fit=640%2C264)
![](https://martinsbioblogg.files.wordpress.com/2019/06/blogpost.png?w=450&h=264&fit=640%2C264)



*Related*







---
