---
layout:     post
catalog: true
title:      Communication Breakdown
subtitle:      转载自：http://feedproxy.google.com/~r/RBloggers/~3/LGCSi5P9U3k/
date:      2019-06-11
author:      quantixed
tags:
    - papers
    - alldf
    - paper titles
    - cell biology
    - citations
---


There is an entertaining rumour going around about the journal *Nature Communications*. When I heard it for the fourth or fifth time, I decided to check out whether there is any truth in it.

> 
The rumour goes something like this: the impact factor of *Nature Communications* is driven by physical sciences papers.


Sometimes it is put another way: cell biology papers drag down the impact factor of *Nature Communications*, or that they don’t deserve the high JIF tag of the journal because they are cited at lower rates. Could this be true?

**TL;DR** it is true but the effect is not as big as the rumour suggests.

*Nature Communications* is the mega-journal that sits below the subject-specific Nature journals. Operating as an open access, pay-to-publish journal it is a way for Springer Nature to capture revenue from papers that were good, but did not make the editorial selection for subject-specific Nature journals. This is a long-winded way of saying that there are wide variety of papers covered by this journal which publishes around 5,000 papers per year. This complicates any citation analysis because we need a way to differentiate papers from different fields. I describe one method to do this below.

#### Quick look at the data

I had a quick look at the top 20 papers from 2016-2017 with the most citations in 2018. There certainly were a lot of non-biological papers in there. Since highly cited papers disproportionately influence the Journal Impact Factor, then this suggested the rumours might be true. 
|Citations (2018)|Title|
|238|11.4% Efficiency non-fullerene polymer solar cells with trialkylsilyl substituted 2D-conjugated polymer as donor|
|226|Circular RNA profiling reveals an abundant circHIPK3 that regulates cell growth by sponging multiple miRNAs|
|208|Recommendations for myeloid-derived suppressor cell nomenclature and characterization standards|
|203|High-efficiency and air-stable P3HT-based polymer solar cells with a new non-fullerene acceptor|
|201|One-Year stable perovskite solar cells by 2D/3D interface engineering|
|201|Massively parallel digital transcriptional profiling of single cells|
|177|Array of nanosheets render ultrafast and high-capacity Na-ion storage by tunable pseudocapacitance|
|166|Multidimensional materials and device architectures for future hybrid energy storage|
|163|Coupled molybdenum carbide and reduced graphene oxide electrocatalysts for efficient hydrogen evolution|
|149|Ti3C2 MXene co-catalyst on metal sulfide photo-absorbers for enhanced visible-light photocatalytic hydrogen production|
|149|Balancing surface adsorption and diffusion of lithium-polysulfides on nonconductive oxides for lithium-sulfur battery design|
|146|Adaptive resistance to therapeutic PD-1 blockade is associated with upregulation of alternative immune checkpoints|
|140|Conductive porous vanadium nitride/graphene composite as chemical anchor of polysulfides for lithium-sulfur batteries|
|136|Fluorination-enabled optimal morphology leads to over 11% efficiency for inverted small-molecule organic solar cells|
|134|The somatic mutation profiles of 2,433 breast cancers refines their genomic and transcriptomic landscapes|
|132|Photothermal therapy with immune-adjuvant nanoparticles together with checkpoint blockade for effective cancer immunotherapy|
|131|Enhanced electronic properties in mesoporous TiO2 via lithium doping for high-efficiency perovskite solar cells|
|125|Electron-phonon coupling in hybrid lead halide perovskites|
|123|A sulfur host based on titanium [email protected] hollow spheres for advanced lithium-sulfur batteries|
|121|Biodegradable black phosphorus-based nanospheres for in vivo photothermal cancer therapy|

#### Let’s dive in to the data

We will use R for this analysis. If you want to work along, the script and data can be downloaded below. With a few edits, the script will also work for similar analysis of other journals.

First of all I retrieved three datasets.

- Citation data for the journal. We’ll look at 2018 Journal Impact Factor, so we need citations in 2018 to papers in the journal published in 2016 and 2017. This can be retrieved from Scopus as a csv.

- Pubmed XML file for the Journal to cover the articles that we want to analyse. Search term = “Nat Commun”[Journal] AND (“2016/01/01″[PDAT] : “2017/12/31″[PDAT])

- Pubmed XML file to get cell biology MeSH terms. Search term = “J Cell Sci”[Journal] AND (“2016/01/01″[PDAT] : “2017/12/31″[PDAT])


#### Using MeSH terms to segregate the dataset

Analysing the citation data is straightforward, but how can we classify the content of the dataset? I realised that I could use Medical Subject Heading (MeSH) from PubMed to classify the data. If I retrieved the same set of papers from PubMed and then check which papers had MeSH terms which matched that of a “biological” dataset, the citation data could be segregated. I used a set of J Cell Sci papers to do this. **Note that these MeSH terms are not restricted to cell biology, they cover all kinds of biochemistry and other aspects of biology.** The papers that do *not* match these MeSH terms are ecology, chemistry and physical sciences (many of these don’t have MeSH terms). We start by getting our biological MeSH terms.

Now we need to load in the Nature Communications XML data from PubMed and also get the citation data into R.

Next we need to perform a *join* to match up the PubMed data with the citation data.

Now we’ll make some plots.

The citation data look typical: highly skewed, with few very highly cited papers and the majority (two-thirds) receiving less than the mean number of citations. The “cell biology” dataset and the non-cell biology dataset look pretty similar.

Now it is time to answer our main question. Do cell biology papers drag down the impact factor of the journal?

The “JIF” for the whole journal is 11.48, whereas for the non-cell biology content it is 13.61. The cell biology dataset has a “JIF” of 10.17. So basically, **the rumour is true but the effect is quite mild**. The rumour is that the cell biology impact factor is *much* lower.

The reason “JIF” is in quotes is that it is notoriously difficult to calculate this metric. All citations are summed for the numerator, but the denominator comprises “citable items”. To get something closer to the actual JIF, we probably should remove non-citable items. These are Errata, Letters, Editorials and Retraction notices.

After removal the citation distributions look a bit more realistic (notice that the earlier versions had many items with zero citations).
![](https://i1.wp.com/quantixed.org/wp-content/uploads/2019/06/p4.png?fit=640%2C320&ssl=1&is-pending-load=1)

![](https://i1.wp.com/quantixed.org/wp-content/uploads/2019/06/p4.png?fit=640%2C320&ssl=1)


Now we can redo the last part.

Now the figures have changed. The “JIF” for the whole journal is 11.63, whereas for the non-cell biology content it would 14.06. The cell biology dataset has a “JIF” of 10.19. To more closely approximate the JIF, we need to do:

This made only a minor change, probably because the dataset is so huge (7239 papers for two years with non-citable items removed). If we were to repeat this on another journal with more front content and fewer papers, this distinction might make a bigger change.

#### Conclusion

So the rumour is true *but* the effect is not as big as people say. There’s a ~17% reduction in potential impact factor by including these papers rather than excluding them. However, these papers comprise ~63% of the corpus and they bring in an estimated revenue to the publisher of $12,000,000 per annum.

**It is definitely not true that these papers are under-performing.** Their citation rates are similar to those in the best journals in the field. Note that citation rates do not necessarily reflect the usefulness of the paper. For one thing they are largely an indicator of the volume of a research field. Anyhow,**next time you hear this rumour for someone, you can set them straight.**

And I nearly managed to write an entire post without mentioning that JIF is a terrible metric, but then you knew that didn’t you?

—

The post title comes from “Communication Breakdown” by the might Led Zeppelin from their debut album. I was really tempted to go with “Dragging Me Down” by Inspiral Carpets, but Communication Breakdown was too good to pass up.


*Related*



---
