---
layout:     post
catalog: true
title:      Using R： plotting the genome on a line
subtitle:      转载自：http://feedproxy.google.com/~r/RBloggers/~3/BJIzktQK6cg/
date:      2019-03-31
author:      mrtnj
tags:
    - contigs
    - on_contig
    - contig_lengths
    - data
    - windows
---





Imagine you want to make a Manhattan-style plot or anything else where you want a series of intervals laid out on one axis after one another. If it’s actually a Manhattan plot you may have a friendly R package that does it for you, but here is how to cobble the plot together ourselves with *ggplot2*.

We start by making some fake data. Here, we have three contigs (this could be your chromosomes, your genomic intervals or whatever) divided into one, two and three windows, respectively. Each window has a value that we’ll put on the y-axis.

We will need to know how long each contig is. In this case, if we assume that the windows cover the whole thing, we can get this from the data. If not, say if the windows don’t go up to the end of the chromosome, we will have to get this data from elsewhere (often some genome assembly metadata). This is also where we can decide in what order we want the contigs.

Now, we need to transform the coordinates on each contig to coordinates on our new axis, where we lay the contings after one another. What we need to do is to add an offset to each point, where the offset is the sum of the lengths of the contigs we’ve layed down before this one. We make a function that takes three arguments: two vectors containing the contig of each point and the position of each point, and also the table of lengths we just made.

Now, we use this to transform the start and end of each window. We also transform the vector of the length of the contigs, so we can use it to add vertical lines between the contigs.

It would be nice to label the x-axis with contig names. One way to do this is to take the coordinates we just made for the vertical lines, add a zero, and shift them one position, like so:

Now it’s time to plot! We add one layer of points for the values on the y-axis, where each point is centered on the middle of the window, followed by a layer of vertical lines at the borders between contigs. Finally, we add our custom x-axis, and also some window dressing.

And this is what we get:

![](https://martinsbioblogg.files.wordpress.com/2019/01/rplot.png?w=456)
![](https://martinsbioblogg.files.wordpress.com/2019/01/rplot.png?w=456)


I’m sure your plot will look more impressive, but you get the idea.


*Related*








---
